control/metrics.go

```diff
@@ -160,23 +189,162 @@ func (m *metricType) Timestamp() time.Time {
+	wkey := getMetricKey(ns)
+
+	// if not exist, match `wkey` with cataloged keys
+	if _, exist := mc.mKeys[wkey]; !exist {
+		mc.addItemToMatchingMap(wkey)
+	}
+
+	// mkeys means matched metrics keys
+	mkeys := mc.mKeys[wkey]
+
+	if len(mkeys) == 0 {
+		return nil, errorMetricNotFound(ns)
+	}
+
+	// convert matched keys to a slice of namespaces
+	nss := convertKeysToNamespaces(mkeys)
+
+	return nss, nil
+}
+
+func convertKeysToNamespaces(keys []string) [][]string {
+	// nss is a slice of slices which holds metrics namespaces
+	nss := [][]string{}
+	for _, key := range keys {
+		ns := getMetricNamespace(key)
+		if len(ns) != 0 {
+			nss = append(nss, ns)
+		}
+	}
+
+	return nss
+}
+
+// addItemToMatchingMap adds `wkey` to matching map (or updates if exists) with corresponding cataloged keys as a content;
+// if this 'wkey' does not match to any cataloged keys, it will be removed from matching map
+func (mc *metricCatalog) addItemToMatchingMap(wkey string) {
+	matchedKeys := []string{}
+
+	// wkey contains `.` which should not be interpreted as regexp tokens, but as a single character
+	exp := strings.Replace(wkey, ".", "[.]", -1)
+
+	// change `*` into regexp `.*` which matches any characters
+	exp = strings.Replace(exp, "*", ".*", -1)
+
+	regex := regexp.MustCompile("^" + exp + "$")
+
+	for _, key := range mc.keys {
+		match := regex.FindStringSubmatch(key)
+
+		if match == nil {
+			continue
+		}
+
+		matchedKeys = appendIfMissing(matchedKeys, key)
+	}
+
+	if len(matchedKeys) == 0 {
+		mc.removeItemFromMatchingMap(wkey)
+	} else {
+		mc.mKeys[wkey] = matchedKeys
+	}
+}
+
+// removeItemFromMatchingMap removes `wkey` from matching map
+func (mc *metricCatalog) removeItemFromMatchingMap(wkey string) {
+	if _, exist := mc.mKeys[wkey]; exist {
+		delete(mc.mKeys, wkey)
+	}
+}
+
+// updateMatchingMap updates the contents of matching map
+func (mc *metricCatalog) updateMatchingMap() {
+	for wkey := range mc.mKeys {
+		// add (or update if exist) item `wkey'
+		mc.addItemToMatchingMap(wkey)
+	}
+}
+
+// removeMatchedKey iterates over all items in the mKey and removes `key` from its content
+func (mc *metricCatalog) removeMatchedKey(key string) {
+	for wkey, mkeys := range mc.mKeys {
+		for index, mkey := range mkeys {
+			if mkey == key {
+				// remove this key from slice
+				mc.mKeys[wkey] = append(mkeys[:index], mkeys[index+1:]...)
+			}
+		}
+		// if no matched key left, remove this item from map
+		if len(mc.mKeys[wkey]) == 0 {
+			mc.removeItemFromMatchingMap(wkey)
+		}
+	}
+}
+
+// validateMetricNamespace validates metric namespace in terms of containing not allowed characters and ending with an asterisk
+func validateMetricNamespace(ns []string) error {
+	name := strings.Join(ns, "")
+
+	for _, chars := range notAllowedChars {
+		for _, ch := range chars {
+			if strings.ContainsAny(name, ch) {
+				return errorMetricContainsNotAllowedChars(ns)
+			}
+		}
+	}
+
+	// todo uncomment this after adjusting all plugins
```

Comment:
Thanks, I decided to remove this commented out code. 

When I finish revising that all plugins do not exposed a wildcard in the end of namespace + refactoring that one which has such like metrics, I will add this part of validation + unit test which covers that again in another pull request.

Already even plugin has metrics which ends with wildcard, it will works properly,  So, as a first step the plugins will be refactored